crates/bevy_pbr/src/render/shadow_sampling.wgsl

```diff
@@ -0,0 +1,156 @@
+fn sample_shadow_map_the_witness(light_local: vec2<f32>, depth: f32, array_index: i32) -> f32 {
+    let shadow_map_size = vec2<f32>(textureDimensions(directional_shadow_textures));
+    let inv_shadow_map_size = 1.0 / shadow_map_size;
+
+    let uv = light_local * shadow_map_size;
+    var base_uv = floor(uv + 0.5);
+    let s = (uv.x + 0.5 - base_uv.x);
+    let t = (uv.y + 0.5 - base_uv.y);
+    base_uv -= 0.5;
+    base_uv *= inv_shadow_map_size;
+
+    let uw0 = (4.0 - 3.0 * s);
+    let uw1 = 7.0;
+    let uw2 = (1.0 + 3.0 * s);
+
+    let u0 = (3.0 - 2.0 * s) / uw0 - 2.0;
+    let u1 = (3.0 + s) / uw1;
+    let u2 = s / uw2 + 2.0;
+
+    let vw0 = (4.0 - 3.0 * t);
+    let vw1 = 7.0;
+    let vw2 = (1.0 + 3.0 * t);
+
+    let v0 = (3.0 - 2.0 * t) / vw0 - 2.0;
+    let v1 = (3.0 + t) / vw1;
+    let v2 = t / vw2 + 2.0;
+
+    var sum = 0.0;
+
+    sum += uw0 * vw0 * sample_shadow_map_simple(base_uv + (vec2(u0, v0) * inv_shadow_map_size), depth, array_index);
+    sum += uw1 * vw0 * sample_shadow_map_simple(base_uv + (vec2(u1, v0) * inv_shadow_map_size), depth, array_index);
+    sum += uw2 * vw0 * sample_shadow_map_simple(base_uv + (vec2(u2, v0) * inv_shadow_map_size), depth, array_index);
+
+    sum += uw0 * vw1 * sample_shadow_map_simple(base_uv + (vec2(u0, v1) * inv_shadow_map_size), depth, array_index);
+    sum += uw1 * vw1 * sample_shadow_map_simple(base_uv + (vec2(u1, v1) * inv_shadow_map_size), depth, array_index);
+    sum += uw2 * vw1 * sample_shadow_map_simple(base_uv + (vec2(u2, v1) * inv_shadow_map_size), depth, array_index);
+
+    sum += uw0 * vw2 * sample_shadow_map_simple(base_uv + (vec2(u0, v2) * inv_shadow_map_size), depth, array_index);
+    sum += uw1 * vw2 * sample_shadow_map_simple(base_uv + (vec2(u1, v2) * inv_shadow_map_size), depth, array_index);
+    sum += uw2 * vw2 * sample_shadow_map_simple(base_uv + (vec2(u2, v2) * inv_shadow_map_size), depth, array_index);
+
+    return sum / 144.0;
+}
+
+// https://blog.demofox.org/2022/01/01/interleaved-gradient-noise-a-different-kind-of-low-discrepancy-sequence
+fn interleaved_gradient_noise(pixel_coordinates: vec2<f32>) -> f32 {
+    let frame = f32(globals.frame_count % 64u);
+    let xy = pixel_coordinates + 5.588238 * frame;
+    return fract(52.9829189 * fract(0.06711056 * xy.x + 0.00583715 * xy.y));
+}
+
+// https://www.iryoku.com/next-generation-post-processing-in-call-of-duty-advanced-warfare (slides 120-135)
+fn sample_shadow_map_stochastic(light_local: vec2<f32>, depth: f32, array_index: i32) -> f32 {
+    let shadow_map_size = vec2<f32>(textureDimensions(directional_shadow_textures));
+    let sample_offsets = array(
+        vec2(-0.7071,  0.7071),
+        vec2(-0.0000, -0.8750),
+        vec2( 0.5303,  0.5303),
+        vec2(-0.6250, -0.0000),
+        vec2( 0.3536, -0.3536),
+        vec2(-0.0000,  0.3750),
+        vec2(-0.1768, -0.1768),
+        vec2( 0.1250,  0.0000),
+    );
```

Comment:
Not sure if it makes any practical difference in a shader but I feel like this should be in a global `var<private>` outside the function.

Reply:
I'd rather keep it inline, so it's clear what function it belongs to, unless you feel very strongly otherwise.

Reply:
I'm not sure if it will reconstruct it on each call to the function. I would hope not. With a `var<private>` at global scope, we could be sure, and code-wide it can be moved just above this function.

Reply:
Naga dosen't support this for arrays atm